### PR TITLE
Don't accept spaces after heredoc beginning identifier. (was: Make Heredoc syntax highlighting work)

### PR DESCRIPTION
**Pull request by Talon1024**
I've modified some of the regular expressions so that Heredoc syntax highlighting now works.

Also, I've disabled the "internal heredocs" because, as far as I know, there is no such thing in PHP.

### CODE REVIEW

**Comment by Ingramz** [comment]
heredoc_interior rule is supposed to provide a semantic way of embedding different sources. While not specifically a language feature, it makes an useful convention to follow and the highlighting is there to help with that.

This works with linguist and textmate's PHP bundle in its current form.

```php
<?php

$a = <<<GITHUB
This is a plain string.
SELECT * FROM github WHERE octocat = 'awesome' and ID = 1;
<strong>rainbows</strong>

if(awesome) {
    doSomething(10, function(x){
      console.log(x*x);
    });
}
GITHUB;

$b = <<<SQL
SELECT * FROM github WHERE octocat = 'awesome' and ID = 1;
SQL;

$c = <<<HTML
<strong>rainbows</strong>
HTML;

$d = <<<JAVASCRIPT
if(awesome) {
    doSomething(10, function(x){
      console.log(x*x);
    });
}
JAVASCRIPT;

$e = <<<JSON
{"menu": {
  "id": "file",
  "value": "File",
  "popup": {
    "menuitem": [
      {"value": "New", "onclick": "CreateNewDoc()"},
      {"value": "Open", "onclick": "OpenDoc()"},
      {"value": "Close", "onclick": "CloseDoc()"}
    ]
  }
}}
JSON;

?>
```

As for the actual issue, I think it has something to do with the lookahead + `(?!\\G)` plainly not working in this case (possibly an issue with first-mate and not with the grammar itself), but right now I can't really comment on that.

**Comment by Talon1024** [comment]
Why were lookahead and lookbehind (AKA lookaround) being used in the first place? In my experience, lookaround does not work well with whitespace.

**Comment by Ingramz** [comment]
The main reason is because textmate grammar had it - why change something that has been proven to work, right?

The actual reason however is because if you match something without lookaround, you cannot match it again. Lookaheads in this case allow to match a generic version of the heredoc/nowdow beginning and then more specific versions can be matched instead inside the block, a nifty way of enabling the functionality for heredoc_interior here. As far as I know the whitespace is no problem. 

What confuses me a little the usage of `(?!\\G)` pattern in combination with the lookahead.

Unfortunately I cannot look into this issue before Saturday.

**Comment by Talon1024** [comment]
Now that I think of it, the point of doing a lookahead for the beginning of the heredoc is to match the position before the heredoc.

I guess the "ending" token should then match the position after the ending identifier.

**Comment by Ingramz** [comment]
Either way for some odd reason right now only the first character from `<<<SOMETHING` is in the heredoc scope and not all of the block.

Pinging @kevinsawicki / @nathansobo for some help with demystifying first-mate behavior in this case.

**Comment by Talon1024** [comment]
I think the pattern `(?!\\G)` will match any position/character that isn't proceded by the beginning of the start of the match. The beginning and end regular expressions are, of course, evaluated in that order, and the ending regular expression can only be evaluated once. Since the first `<` sign isn't proceded by the start of the match (it appears after the start of the match), it is the only character in the heredoc scope.

**Comment by Ingramz** [comment]
Hmm... after sleeping on it, you are right.

**Comment by Ingramz** [comment]
Still I do not believe that changing the grammar is the solution here. We are talking of at least 3 implementations where this exact regex works and now it just doesn't.

**Comment by nathansobo** [comment]
I probably broke this with some changes to the parser intended to save memory. I will look into it Monday at the latest.

**Comment by Ingramz** [comment]
@nathansobo it's possible that heredoc strings have never worked in Atom.

**Comment by Talon1024** [comment]
[Heredoc syntax highlighting has been broken since Atom 0.174](https://github.com/atom/language-php/issues/57).

**Comment by Ingramz** [comment]
@Talon1024 Thanks, my bad.

**Comment by Ingramz** [comment]
@Talon1024 0.174 did not break it, so the search for a working version continues.

**Comment by Talon1024** [comment]
I just tried Atom 0.45, and heredoc syntax highlighting didn't work in that version either. On the other hand, maybe try matching as much as possible in between the beginning and end tokens instead of as little as possible... Just a thought.

**Comment by Talon1024** [comment]
So, I've compiled my own build of TextMate using the source code from their GitHub repo, and I've played around with the TextMate PHP bundle using the built-in bundle editor.

It works properly if no changes are made to the bundle.

If I remove the `#heredoc_interior` include from the TextMate grammar, the highlighting breaks, and the code ends up being highlighted in the same way it is in Atom now (only the first < character is in the heredoc scope)

It is partially fixed, with no syntax highlighting for SQL, HTML, JS, or JSON if I make the equivalent changes that I made to the TextMate bundle.

These findings have led me to believe that "includes" in TextMate work differently than they do in Atom. I think the way it works in TextMate is that, if a match for the included grammar element is found at the starting position of the parent grammar element, it will highlight from that point using the included grammar element regexes.

However, [The grammar for this package has been changed in the past](#73), and it's ultimately up to whoever is in charge of Atom and this bundle if they want 1:1 TextMate bundle compatibility.

**Comment by 50Wliu** [comment]
This no longer merges cleanly; @Talon1024 would you mind rebasing?

**Comment by Talon1024** [comment]
OK, I've changed my repo to be up to date with this repo, and I redid my changes from scratch.

**Comment by 50Wliu** [comment]
I'll try to review this tonight.

**Comment by Ingramz** [comment]
Not that I'd like to see this to be merged, I believe that this workaround won't motivate to solve the underlying cause in first-mate or underlying regular expression engine. I hope that merging this right now won't do that.

Also someone should open an issue in first-mate for that (I can do it later tonight).

**Comment by Talon1024** [comment]
Oh my goodness. I was able to fix this without changing any of the expressions, but simply moving the `heredoc_interior` rule out of a nested `repository`.

Take a look at the structure of the JSON generated from php.cson:
phpgrammarjsonstructure

Notice the `heredoc_interior` rule is in a nested `repository`? If I move it out of the nested `repository`, heredocs will highlight correctly.
phpgrammarhackfixed

So, the problem is really first-mate ignoring nested `repositories`.

**Comment by Ingramz** [comment]
This seems awfully familiar when I tried to fix it. I ended up posting this: https://github.com/atom/first-mate/issues/38#issuecomment-138676408

**Comment by Talon1024** [comment]
In this commit, I've restored the original structure of php.cson, but I've modified the expressions so that they won't accept spaces after the heredoc beginning identifier, since that will cause a PHP syntax error.

**Comment by Ingramz** [comment]
LGTM :+1:

**Comment by 50Wliu** [comment]
Please add tests to this :smile:.

**Comment by Ingramz** [comment]
Also note that we don't want to highlight newline characters incorrectly, which is what capture group `(\s*) ` does, although it's fine for the case when you aren't highlighting those characters. That is why I suggested  `([ \t]*)` instead. Atom doesn't have highlighting support for end-of-lines right now (although it still matches them), but in order to keep the grammar accurate, I suggest you updating all the corresponding capturing groups to match space and tab only.

**Comment by Talon1024** [comment]
Keep in mind that `$` is an anchor that matches the end of the line, so I don't think (\s*) will match a newline if it comes before a `$`, which it does.

**Comment by Talon1024** [comment]
I just added some breaking tests and modified the grammar so that a zero-length string at the end of the line will not get tokenized as invalid/illegal.

**Comment by Ener-Getick** [comment]
@Talon1024 did you stop working on this ?
It would be amazing to this working and merged !

**Comment by Aldlevine** [comment]
Any progress on this? It would be an awesome feature to have! :+1:

**Comment by Wardevour** [comment]
I have a block like:
```php
$tst = <<<HTML
<script type="text/javascript">
try {
    // code goes here...
}
catch(e) {
    alert(e.message);
}
</script>
HTML;

```

JS in a heredoc is probably frowned upon, but the syntax highlighting stops working on line 6 because it's not a correct PHP catch block (missing Exception class type). Will this PR fix my issue or am I having a different problem?

**Comment by UziTech** [comment]
> Notice the heredoc_interior rule is in a nested repository? If I move it out of the nested repository, heredocs will highlight correctly.

If there was a fix for this over a year ago why hasn't this been fixed yet? I realize it would only be a temporary but it sure would've beaten no fix while we wait on an issue that some other package [isn't even working on.](https://github.com/atom/first-mate/issues/38#issuecomment-141123296)

Can we please just get a temporary fix done while we wait on this pull request?

**Comment by 50Wliu** [comment]
I would much rather the underlying issue be fixed than piling on "temporary" workarounds, because from experience it then becomes "oh, there's a workaround, so we can deprioritize fixing the actual issue" and the workaround becomes not-so-temporary.

At any rate, I can't merge this even if I wanted to because it has conflicts.

**Comment by UziTech** [comment]
If it is between no fix for a year+ (and god knows how much longer) or a temporary fix, I would rather have the temporary fix. Especially since [this](https://github.com/atom/language-php/issues/10) [obviously](https://github.com/atom/language-php/issues/57) [affects](https://github.com/atom/language-php/issues/63) [a](https://github.com/atom/language-php/issues/78) [lot](https://github.com/atom/language-php/issues/113) [of](https://github.com/atom/language-php/issues/133) [people](https://github.com/atom/language-php/issues/155) 

If the underlying issue was a priority, I would agree with you. But it is obviously not.

I created a new pull request #184

File: spec/php-spec.coffee [hunk 1]
```diff
@@ -437,3 +437,151 @@ describe 'PHP grammar', ->
...
+        expect(tokens[1][3]).toEqual value: '=', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'keyword.operator.assignment.php']
+        expect(tokens[1][4]).toEqual value: ' ', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php']
+        expect(tokens[1][5]).toEqual value: '<<<', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.php', 'punctuation.definition.string.php']
+        expect(tokens[1][6]).toEqual value: 'HEREDOC', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.php', 'keyword.operator.heredoc.php']
+        expect(tokens[2][0]).toEqual value: 'I am a heredoc', scopes ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.php']
+        expect(tokens[3][0]).toEqual value: 'HEREDOC', scopes ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.php', 'keyword.operator.heredoc.php']
+        expect(tokens[3][1]).toEqual value: ';', scopes ['text.html.php', 'meta.embedded.block.php', 'source.php', 'punctuation.terminator.expression.php']
```
Comment (UziTech) [inline spec/php-spec.coffee]: lines 454 - 456

There should be a colon (:) after `scopes`

**Comment by 50Wliu** [comment]
Superseded by #184.

**Pull request closed without merging**